File: rec-core/src/main/java/net/kimleo/rec/repository/selector/Selector.java
package net.kimleo.rec.repository.selector;

import net.kimleo.rec.repository.RecRepository;
import net.kimleo.rec.repository.RecordSet;
import net.kimleo.rec.repository.selector.expr.SelectorParser;

import java.util.List;

public interface Selector {
    List<RecordSet> findAll(RecRepository repo);


    static Selector of(String expr) {
        List<Selector> selectors = SelectorParser.INSTANCE.parse(SelectorParser.INSTANCE.lex(expr));
        return new CombinedSelector(selectors);
    }
}

File: rec-core/src/main/kotlin/net/kimleo/rec/repository/selector/Selector.kt
package net.kimleo.rec.repository.selector

import net.kimleo.rec.repository.RecCollection
import net.kimleo.rec.repository.RecRepository
import net.kimleo.rec.repository.selector.expr.SelectorParser

interface Selector {
    fun findAll(repo: RecRepository): List<RecCollection>

    companion object {
        fun of(expr: String): Selector {
            val selectors = SelectorParser.parse(SelectorParser.lex(expr))
            return CombinedSelector(selectors)
        }
    }
}